packages/components/src/components/Tabs/hooks/useTabTriggers.tsx

```diff
@@ -0,0 +1,139 @@
+    activeIndicatorRef: RefObject<HTMLSpanElement>,
+) => {
+    const activeTriggerElement = triggerListRef.current?.querySelector('[data-state="active"]') as HTMLButtonElement;
+    const triggerListElement = triggerListRef.current as HTMLDivElement;
+    const activeIndicatorElement = activeIndicatorRef.current as HTMLSpanElement;
+
+    if (!triggerListElement || !activeTriggerElement) {
+        return;
+    }
+
+    const isOverflowingLeft = triggerListElement?.scrollLeft > activeTriggerElement?.offsetLeft;
+    const isOverflowingRight =
+        activeTriggerElement?.offsetLeft + activeTriggerElement?.offsetWidth > triggerListElement?.offsetWidth;
+
+    if (isOverflowingLeft) {
+        const willFitAllItemsToTheLeft =
+            triggerListElement.offsetWidth > activeTriggerElement.offsetWidth + activeTriggerElement.offsetLeft;
+        if (willFitAllItemsToTheLeft) {
+            activeIndicatorElement.style.left = `${activeTriggerElement.offsetLeft}px`;
+        } else {
+            activeIndicatorElement.style.left = `${triggerListElement.offsetWidth - activeTriggerElement.offsetWidth}px`;
+        }
+        activeIndicatorElement.style.width = `${activeTriggerElement?.offsetWidth}px`;
+    } else if (isOverflowingRight) {
+        activeIndicatorElement.style.left = `${triggerListElement.clientWidth - activeTriggerElement.offsetWidth}px`;
+        activeIndicatorElement.style.width = `${activeTriggerElement.offsetWidth}px`;
+    } else {
+        activeIndicatorElement.style.left = `${activeTriggerElement.offsetLeft}px`;
+        activeIndicatorElement.style.width = `${activeTriggerElement.offsetWidth}px`;
+    }
+};
+
+export const useTabTriggers = ({
+    activeTab,
+}: {
+    activeTab?: string;
+}): {
+    triggerListRef: RefObject<HTMLDivElement>;
+    activeIndicatorRef: RefObject<HTMLSpanElement>;
+    triggers: TabTrigger[];
+    triggersOutOfView: TabTrigger[];
+    addTrigger: (trigger: TabTrigger) => void;
+} => {
+    const triggerListRef = useRef<HTMLDivElement>(null);
+    const activeIndicatorRef = useRef<HTMLSpanElement>(null);
+    const [triggers, setTriggers] = useState<TabTrigger[]>([]);
+    const [triggersOutOfView, setTriggersOutOfView] = useState<TabTrigger[]>([]);
+
+    // move the active indicator and scroll to the correct position when the tab changes
+    useEffect(() => {
+        const activeTriggerElement = triggerListRef.current?.querySelector('[data-state="active"]');
+        if (activeTriggerElement) {
+            moveActiveIndicator(triggerListRef, activeIndicatorRef);
+            activeTriggerElement.scrollIntoView({ behavior: 'smooth', block: 'nearest', inline: 'end' });
+        }
+    }, [activeTab, activeIndicatorRef]);
+
+    useLayoutEffect(() => {
+        const activeTriggerElement = triggerListRef.current?.querySelector('[data-state="active"]') as HTMLElement;
+        const triggerListElement = triggerListRef.current as HTMLDivElement;
```

Comment:
Use proper runtime assertion
`if (activeTriggerElement && triggerListElement)` (and even some `instanceof`)

Because it could fail to find with `querySelector` in case:
- there is no `data-state`
- `triggerListRef.current` is `null`

```suggestion
        const activeTriggerElement = triggerListRef.current?.querySelector('[data-state="active"]');
        const triggerListElement = triggerListRef.current;
```

Reply:
added runtime assertion